core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnTableScan.scala

```diff
@@ -63,22 +65,109 @@ import org.apache.spark.{Dependency, Partition, RangeDependency, SparkContext, T
+
+    // Returned filter predicate should return false iff it is impossible for the input expression
+    // to evaluate to `true' based on statistics collected about this partition batch.
+    // This code is picked up from InMemoryTableScanExec
+    @transient def buildFilter: PartialFunction[Expression, Expression] = {
+      case And(lhs: Expression, rhs: Expression)
+        if buildFilter.isDefinedAt(lhs) || buildFilter.isDefinedAt(rhs) =>
+        (buildFilter.lift(lhs) ++ buildFilter.lift(rhs)).reduce(_ && _)
+
+      case Or(lhs: Expression, rhs: Expression)
+        if buildFilter.isDefinedAt(lhs) && buildFilter.isDefinedAt(rhs) =>
+        buildFilter(lhs) || buildFilter(rhs)
+
+      case EqualTo(a: AttributeReference, l: Literal) =>
+        statsFor(a).lowerBound <= l && l <= statsFor(a).upperBound
+      case EqualTo(l: Literal, a: AttributeReference) =>
+        statsFor(a).lowerBound <= l && l <= statsFor(a).upperBound
+
+      case LessThan(a: AttributeReference, l: Literal) => statsFor(a).lowerBound < l
+      case LessThan(l: Literal, a: AttributeReference) => l < statsFor(a).upperBound
+
+      case LessThanOrEqual(a: AttributeReference, l: Literal) => statsFor(a).lowerBound <= l
+      case LessThanOrEqual(l: Literal, a: AttributeReference) => l <= statsFor(a).upperBound
+
+      case GreaterThan(a: AttributeReference, l: Literal) => l < statsFor(a).upperBound
+      case GreaterThan(l: Literal, a: AttributeReference) => statsFor(a).lowerBound < l
+
+      case GreaterThanOrEqual(a: AttributeReference, l: Literal) => l <= statsFor(a).upperBound
+      case GreaterThanOrEqual(l: Literal, a: AttributeReference) => statsFor(a).lowerBound <= l
+
+      case IsNull(a: Attribute) => statsFor(a).nullCount > 0
+      case IsNotNull(a: Attribute) => statsFor(a).count - statsFor(a).nullCount > 0
+    }
+
+    // This code is picked up from InMemoryTableScanExec
+    val cachedBatchStatFilters: Seq[Expression] = {
+      if (relation.isInstanceOf[BaseColumnFormatRelation]) {
+        allFilters.flatMap { p =>
+          val filter = buildFilter.lift(p)
+          val boundFilter =
+            filter.map(
+              BindReferences.bindReference(
+                _,
+                cachedBatchStatistics.schema,
+                allowFailures = true))
+
+          boundFilter.foreach(_ =>
+            filter.foreach(f => logInfo(s"Predicate $p generates partition filter: $f")))
+
+          // If the filter can't be resolved th en we are missing required statistics.
+          boundFilter.filter(_.resolved)
+        }
+      } else {
+        Seq.empty[Expression]
+      }
+    }
+
+    metricsCreatedBeforeInit = metricsCreatedBeforeInit ++
+        Map("cachedBatchesSeen" -> SQLMetrics.createMetric(sparkContext,
+      "cached batches seen"),
+    "cachedBatchesSkipped" -> SQLMetrics.createMetric(sparkContext,
+      "cached batches skipped by the predicate"))
+
+    new StatsPredicate(newPredicate,
```

Comment:
Can you please rename it to StatsPredicateCompiler ? It will suggest we are passing a function rather than compiled predicates.